portal-impl/src/com/liferay/portal/util/PortalClassPathUtil.java

```diff
@@ -189,142 +190,178 @@ private static String _buildClassPath(
 		return sb.toString();
 	}
 
-	private static File[] _listClassPathFiles(
-		ClassLoader classLoader, String className) {
+	private static boolean _isPetraJar(File file) {
+		String filePath = file.getAbsolutePath();
+
+		if (filePath.contains("petra")) {
+			try (JarFile jarFile = new JarFile(new File(filePath))) {
+				Manifest manifest = jarFile.getManifest();
 
-		String pathOfClass = StringUtil.replace(
-			className, CharPool.PERIOD, CharPool.SLASH);
+				if (manifest == null) {
+					return false;
+				}
 
-		pathOfClass = pathOfClass.concat(".class");
+				Attributes attributes = manifest.getMainAttributes();
 
-		URL url = classLoader.getResource(pathOfClass);
+				if (attributes.containsKey("Liferay-Releng-App-Title")) {
+					return false;
+				}
 
-		if (_log.isDebugEnabled()) {
-			_log.debug("Build class path from " + url);
+				return true;
+			}
+			catch (IOException ioException) {
+				_log.error(
+					"Unable to resolve bootstrap entry: " + file.getName() +
+						" from bundle",
+					ioException);
+			}
 		}
 
-		String protocol = url.getProtocol();
+		return false;
+	}
 
-		if (protocol.equals("bundle") || protocol.equals("bundleresource")) {
-			try {
-				URLConnection urlConnection = url.openConnection();
+	private static List<File> _listClassPathFiles(
```

Comment:
@julschong Do you have a better solution to fix the issue I mentioned without modifying _listClassPathFiles? Please slack me tomorrow and we can have a talk.

Tina.